source/ui/memory-component-presenter.cpp

```diff
@@ -100,58 +121,126 @@ MemoryComponentPresenter::data(const QModelIndex &index, int role) const {
+    //  stringvalue = StringConversions::toOctString(memory_cell);
+    else if (role_string.startsWith("hex")) {
+      memoryStringValue = StringConversions::toHexString(memory_cell);
+    } else if (role_string.startsWith("decs")) {
+      memoryStringValue = StringConversions::toSignedDecString(memory_cell);
+    } else if (role_string.startsWith("dec")) {
+      memoryStringValue = StringConversions::toUnsignedDecString(memory_cell);
+    } else {
+      memoryStringValue = "unknown format";
+    }
+
+    return QString::fromStdString(memoryStringValue);
+  } else {
+    // wrong address
     return QString("");
   }
+}
 
-  int memory_address = index.row() * (number_of_bits / 8);
-  int memory_length = number_of_bits / 8;
-  MemoryValue memory_cell =
-      _memoryAccess.getMemoryValueAt(memory_address, memory_length).get();
 
-  std::string stringValue;
-  if (role_string.startsWith("bin")) {
-    stringValue = StringConversions::toBinString(memory_cell);
-  }
-  // not yet implemented by conversions
-  // else if (role_string.startsWith("oct"))
-  //  stringvalue = StringConversions::toOctString(memory_cell);
-  else if (role_string.startsWith("hex")) {
-    stringValue = StringConversions::toHexString(memory_cell);
-  } else if (role_string.startsWith("decs")) {
-    stringValue = StringConversions::toSignedDecString(memory_cell);
-  } else if (role_string.startsWith("dec")) {
-    stringValue = StringConversions::toUnsignedDecString(memory_cell);
+QVariant
+MemoryComponentPresenter::dataInfo(const QModelIndex &index, int role) const {
+  // TODO fetch value from core
+  return QString("");
+}
+
+
+MemoryValue MemoryComponentPresenter::getMemoryValueCached(
+    const std::size_t memory_address, const std::size_t memory_length) const {
+  // check for cache
+  if (_memoryCacheValid && memory_address >= _memoryCacheBaseAddress &&
+      memory_address + memory_length <=
+          _memoryCacheBaseAddress + _memoryCacheSize) {
+    // cache hit -> get memory value from cache
+
+    // fetch data from cache
+    return _memoryCache.subSet(
+        (memory_address - _memoryCacheBaseAddress) * 8,
+        (memory_address - _memoryCacheBaseAddress + memory_length) * 8);
   } else {
-    stringValue = "unknown format";
-  }
+    // cache miss -> update cache from core
 
-  return QString::fromStdString(stringValue);
+    // calculate new cache region
+    _memoryCacheBaseAddress = memory_address;
+    _memoryCacheSize = memory_length;
+
+    // add some offset around but
+    // cache size should not exceed real memory boundaries
+    // (done this way to prevent overflows)
+    const int offset = 10;
+    if (_memoryCacheBaseAddress < offset)
+      _memoryCacheBaseAddress = 0;
+    else
+      _memoryCacheBaseAddress -= offset;
+    if (_memoryCacheBaseAddress + _memoryCacheSize >= _memorySize - offset * 2)
+      _memoryCacheSize = _memorySize - _memoryCacheBaseAddress;
+    else
+      _memoryCacheSize += offset * 2;
```

Comment:
`std::max` can not be used in the first assignment because it will compare 0 with a very large number due to an overflow. You are right that `std::min` can be used in the second case. But personally I would like to prefer to keep both cases similar.